index.js

```diff
@@ -1,9 +1,93 @@
 const daggy = require('daggy')
 const Dict = daggy.tagged('data')
+const Option = require('fantasy-options')
+const Either = require('fantasy-eithers')
+const Identity = require('fantasy-identities')
+
 module.exports = Dict
 
+// Helper functions (shamelessly ripoff from https://github.com/component/clone)
+// slight modification to have clones share their prototype
+// [TODO] submit a Pull Request to original project and use it as an external dependency for object cloning ?
+function type (val) {
+  switch (toString.call(val)) {
+    case '[object Date]': return 'date'
+    case '[object RegExp]': return 'regexp'
+    case '[object Arguments]': return 'arguments'
+    case '[object Array]': return 'array'
+    case '[object Error]': return 'error'
+  }
+  if (val === null) return 'null'
+  if (val === undefined) return 'undefined'
+//  if (val !== val) return 'nan'
+  if (val && val.nodeType === 1) return 'element'
+
+  val = val.valueOf
+    ? val.valueOf()
+    : Object.prototype.valueOf.apply(val)
+
+  return typeof val
+}
+
+function clone (obj) {
+  switch (type(obj)) {
+    case 'object':
+
+      // preserve prototype of cloned object
+      copy = Object.create(Object.getPrototypeOf(obj))
+
+      for (var key in obj) {
+        if (Object.prototype.hasOwnProperty.call(obj, key)) {
+          copy[key] = clone(obj[key])
+        }
+      }
+      return copy
+
+    case 'array':
+      var copy = new Array(obj.length)
+      for (var i = 0, l = obj.length; i < l; i++) {
+        copy[i] = clone(obj[i])
+      }
+      return copy
+
+    case 'date':
+      return new Date(obj.getTime())
+
+    default: // string, number, boolean, …
+      return obj
+  }
+}
+
+// Fantasy Land has no interface to test if an object is a fantasy-land value
+//   - isADT should provide compatibility with most used Fantasy-land compatible libs (folktale, Sanctuary, Ramda-fantasy, etc.)
+//   - current implementation tests Maybe and Either Monads and should be conservatively extended to to other usefull monads
+const isADT = (obj) =>
+       obj.isJust ||
+       obj.isNothing ||
+       obj.isLeft ||
+       obj.isRight ||
+       obj instanceof Option ||
+       obj instanceof Either ||
+       obj instanceof Identity ||
+       (obj.isJust && (typeof obj.isJust === 'function') && obj.isJust()) ||
+       (obj.isNothing && (typeof obj.isNothing === 'function') && obj.isNothing()) ||
+       (obj.isRight && (typeof obj.isRight === 'function') && obj.isRight()) ||
+       (obj.isLeft && (typeof obj.isLeft === 'function') && obj.isLeft())
+
 // --- Fantasy-Land methods
 
+// Setoid
```

Comment:
I like this Setoid implementation, delegating the work to the `.equals()` method of its children (if it exists).